File: android/app/src/main/java/com/gamebooster/app/device/NetworkStateObserver.java
package com.gamebooster.app.device;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.util.Log;

import androidx.annotation.NonNull;

import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * NetworkStateObserver — Clean, event-driven network connectivity monitor.
 *
 * Requirements addressed:
 *  - Background & Offline Reliability: Shizuku does not require internet; this observer
 *    only notifies listeners of network changes for telemetry/dashboard display.
 *  - Does NOT continuously poll the network or battery.
 *  - Uses modern ConnectivityManager.NetworkCallback.
 */
public class NetworkStateObserver {

    public interface Listener {
        void onNetworkStateChanged(boolean isOnline);
    }

    private static final String TAG = "NetworkStateObserver";

    private static volatile NetworkStateObserver sInstance;

    public static NetworkStateObserver getInstance(Context context) {
        if (sInstance == null) {
            synchronized (NetworkStateObserver.class) {
                if (sInstance == null) {
                    sInstance = new NetworkStateObserver(context);
                }
            }
        }
        return sInstance;
    }

    private final Context appContext;
    private final ConnectivityManager connectivityManager;
    private final CopyOnWriteArrayList<Listener> listeners = new CopyOnWriteArrayList<>();
    private final AtomicBoolean isRegistered = new AtomicBoolean(false);

    private volatile boolean isOnline = false;

    private final ConnectivityManager.NetworkCallback networkCallback = new ConnectivityManager.NetworkCallback() {
        @Override
        public void onAvailable(@NonNull Network network) {
            evaluateConnectivity();
        }

        @Override
        public void onLost(@NonNull Network network) {
            evaluateConnectivity();
        }

        @Override
        public void onCapabilitiesChanged(@NonNull Network network, @NonNull NetworkCapabilities networkCapabilities) {
            evaluateConnectivity();
        }
    };

    private NetworkStateObserver(Context context) {
        this.appContext = context.getApplicationContext();
        this.connectivityManager = (ConnectivityManager) appContext.getSystemService(Context.CONNECTIVITY_SERVICE);
        evaluateConnectivity();
        register();
    }

    public boolean isOnline() {
        return isOnline;
    }

    public void addListener(Listener listener) {
        if (listener != null && !listeners.contains(listener)) {
            listeners.add(listener);
            listener.onNetworkStateChanged(isOnline);
        }
    }

    public void removeListener(Listener listener) {
        if (listener != null) {
            listeners.remove(listener);
        }
    }

    public void register() {
        if (isRegistered.compareAndSet(false, true)) {
            try {
                if (connectivityManager != null) {
                    NetworkRequest request = new NetworkRequest.Builder()
                            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                            .build();
                    connectivityManager.registerNetworkCallback(request, networkCallback);
                    Log.d(TAG, "Registered NetworkStateObserver callback.");
                }
            } catch (Exception e) {
                Log.w(TAG, "Failed to register network callback: " + e.getMessage());
            }
        }
    }

    public void unregister() {
        if (isRegistered.compareAndSet(true, false)) {
            try {
                if (connectivityManager != null) {
                    connectivityManager.unregisterNetworkCallback(networkCallback);
                    Log.d(TAG, "Unregistered NetworkStateObserver callback.");
                }
            } catch (Exception e) {
                Log.w(TAG, "Failed to unregister network callback: " + e.getMessage());
            }
        }
    }

    public boolean evaluateConnectivity() {
        if (connectivityManager == null) return false;
        try {
            Network activeNet = connectivityManager.getActiveNetwork();
            NetworkCapabilities caps = connectivityManager.getNetworkCapabilities(activeNet);
            boolean online = caps != null && (
                    caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
                    caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
            );
            boolean changed = (online != isOnline);
            isOnline = online;
            if (changed) {
                Log.i(TAG, "Network state updated: online=" + isOnline);
                for (Listener listener : listeners) {
                    try {
                        listener.onNetworkStateChanged(isOnline);
                    } catch (Throwable t) {
                        Log.w(TAG, "Error notifying network listener", t);
                    }
                }
            }
        } catch (Throwable t) {
            Log.w(TAG, "Error evaluating connectivity: " + t.getMessage());
        }
        return isOnline;
    }
}

File: android/app/src/main/java/com/gamebooster/app/device/NetworkStateObserver.kt
package com.gamebooster.app.device

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Build
import android.util.Log
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicBoolean

/**
 * NetworkStateObserver — Clean, event-driven network connectivity monitor.
 *
 * Requirements addressed:
 *  - Background & Offline Reliability: Shizuku does not require internet; this observer
 *    only notifies listeners of network changes for telemetry/dashboard display.
 *  - Does NOT continuously poll the network or battery.
 *  - Uses modern ConnectivityManager.NetworkCallback.
 */
class NetworkStateObserver private constructor(context: Context) {

    interface Listener {
        fun onNetworkStateChanged(isOnline: Boolean)
    }

    private val appContext = context.applicationContext
    private val connectivityManager = appContext.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
    private val listeners = CopyOnWriteArrayList<Listener>()
    private val isRegistered = AtomicBoolean(false)

    @Volatile
    var isOnline: Boolean = false
        private set

    private val networkCallback = object : ConnectivityManager.NetworkCallback() {
        override fun onAvailable(network: Network) {
            evaluateConnectivity()
        }

        override fun onLost(network: Network) {
            evaluateConnectivity()
        }

        override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
            evaluateConnectivity()
        }
    }

    init {
        evaluateConnectivity()
        register()
    }

    fun addListener(listener: Listener) {
        if (!listeners.contains(listener)) {
            listeners.add(listener)
            listener.onNetworkStateChanged(isOnline)
        }
    }

    fun removeListener(listener: Listener) {
        listeners.remove(listener)
    }

    fun register() {
        if (isRegistered.compareAndSet(false, true)) {
            try {
                val request = NetworkRequest.Builder()
                    .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                    .build()
                connectivityManager?.registerNetworkCallback(request, networkCallback)
                Log.d(TAG, "Registered NetworkStateObserver callback.")
            } catch (e: Exception) {
                Log.w(TAG, "Failed to register network callback: ${e.message}")
            }
        }
    }

    fun unregister() {
        if (isRegistered.compareAndSet(true, false)) {
            try {
                connectivityManager?.unregisterNetworkCallback(networkCallback)
                Log.d(TAG, "Unregistered NetworkStateObserver callback.")
            } catch (e: Exception) {
                Log.w(TAG, "Failed to unregister network callback: ${e.message}")
            }
        }
    }

    fun evaluateConnectivity(): Boolean {
        val cm = connectivityManager ?: return false
        val activeNet = cm.activeNetwork
        val caps = cm.getNetworkCapabilities(activeNet)
        val online = caps != null && (
                caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
                caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
        )
        val changed = (online != isOnline)
        isOnline = online
        if (changed) {
            Log.i(TAG, "Network state updated: online=$isOnline")
            for (listener in listeners) {
                try {
                    listener.onNetworkStateChanged(isOnline)
                } catch (t: Throwable) {
                    Log.w(TAG, "Error notifying network listener", t)
                }
            }
        }
        return isOnline
    }

    companion object {
        private const val TAG = "NetworkStateObserver"

        @Volatile
        private var instance: NetworkStateObserver? = null

        @JvmStatic
        fun getInstance(context: Context): NetworkStateObserver {
            return instance ?: synchronized(this) {
                instance ?: NetworkStateObserver(context).also { instance = it }
            }
        }
    }
}
